Automated Regression Phase 1

dashattuck opened the pull request:
# Critical Changes

# Changes

## General Updates
We've made some changes to ensure that our regression tests use EDA tests instead of HEDA tests. We've also included several new test automation scripts.

# Issues Closed

# New Metadata

# Deleted Metadata

# Testing Notes

dashattuck commented:
@jonathanjayaraj  - can you review and merge if approved?  Thanks!

robertbwatson commented:
@dashattuck @jonathanjayaraj We might want to consider waiting on this until #768 is merged since there are a lot of robot changes there, including many file renames. Either way, we are likely to run into merge conflicts no matter what order these PRs are merged.

dashattuck commented:
Bobby,

Good call. 

Let’s hold off on this. I’ll re-work these changes after the merge of #768

Doug 


> On Apr 23, 2019, at 12:17 PM, Robert Watson <notifications@github.com> wrote:
> 
> @dashattuck @jonathanjayaraj We might want to consider waiting on this until #768 is merged since there are a lot of robot changes there, including many file renames. Either way, we are likely to run into merge conflicts no matter what order these PRs are merged.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

robertbwatson commented:
> Bobby, Good call. Let’s hold off on this. I’ll re-work these changes after the merge of #768 

Sounds good, @dashattuck! In the meantime, it seems as though the robot tests aren't passing here currently.

lifewithryan commented:
@dashattuck @robertbwatson @jonathanjayaraj -- looks like #768 has been merged. Are we ready to get this one reviewed and merged as well?

lifewithryan commented:
@dashattuck -- what is the status with this PR?

dashattuck commented:
@lifewithryan  - hoping to complete this today.  I've identified the last good checkin - and I'm re-merging the EDA name changes with that good checkin

satishshub commented:
@dashattuck were you able to resolve?

dashattuck commented:
All edits are now complete.   Creating scratch org and testing the results

dashattuck commented:
I need to get with @jstvz and investigate possible bug in EDA path names.  Will do first thing Tuesday morning.

dashattuck commented:
@ceiroa - I've been working on this today.  I made some code changes and am in the process of re-building.  I'm trying to get it resolved today.

dashattuck commented:
Test notes updated.  Ready for merge.

dashattuck commented:
@jstvz - this is ready for merge by you or RelEng team.

robertbwatson commented:
@juliesnow I added you as a reviewer here since there are release notes.

dashattuck commented:
It served the purpose of populating contacts for future test scenarios. 

They may be pulled out for now. 

Doug



> On May 29, 2019, at 4:07 PM, James Estevez <notifications@github.com> wrote:
> 
> @jstvz requested changes on this pull request.
> 
> I had a couple of questions, and made some suggested edits to remove dead code.
> 
> In robot/EDA/tests/browser/contacts/populate_contacts.robot:
> 
> > @@ -0,0 +1,100 @@
> +*** Settings ***
> Curious why we need 10 of these? How are they different? If they aren't different, what do the additional 8 test cases do that the first two don't?
> 
> In robot/EDA/tests/browser/regression/run_regression_suite1.robot:
> 
> > + 
> +    
> +Create Contact And Verify
> +    ${contact_id} =  API Populate Create And Return Contact with Address    Joe      Mazzocco   5345 Calero Ave             San Jose     95023     CA     USA    
> +    &{contact} =  Salesforce Get    Contact             ${contact_id}
> +    Header Field Value              Account Name        &{contact}[LastName] Administrative Account
> +    Select Tab    Details 
> +    Page Should Contain    5345 Calero Ave  
> +    Go To Object Home         Contact
> +    Verify Record    &{contact}[FirstName] &{contact}[LastName]
> +
> +Verify EDA Settings
> +    Go To Eda Settings
> +
> +    Wait for Locator                            eda_settings.affiliations_tab
> +	Click on Element                            eda_settings.affiliations_tab
> ⬇️ Suggested change
> -	Click on Element                            eda_settings.affiliations_tab
> +    Click on Element                            eda_settings.affiliations_tab
> In robot/EDA/tests/browser/regression/run_regression_suite1.robot:
> 
> > +    Click on Element                        course_offering.new_button 
> +
> +    Wait for Locator                        course_offering.new_course_button
> +    Click on Element                        course_offering.new_course_button
> +
> +    Populate Field                          Course Name     robotTest Course Name Test1 - this is just another robot test string in place of a course name
> +    Populate Field                          Department      robotTestLastName Administrative Account
> +    Press Keys                              //input[@title='Search Accounts']          ARROW_DOWN+RETURN
> +    #Press Keys                              TAB 
> +    
> +
> +    Click Element                           (//span[@class=' label bBody' and text()='Save']/ancestor::button[contains(@class, 'slds-button')])[3]
> +    #Click on Element                        course_offering.save_button
> +
> +
> +#    Click Modal Button                      Save
> ⬇️ Suggested change
> -#    Click Modal Button                      Save
> In robot/EDA/tests/browser/regression/run_regression_suite1.robot:
> 
> > +    
> +
> +    Click Element                           (//span[@class=' label bBody' and text()='Save']/ancestor::button[contains(@class, 'slds-button')])[3]
> +    #Click on Element                        course_offering.save_button
> +
> +
> +#    Click Modal Button                      Save
> +    Click on Element                        term.search_terms
> +    Click on Element                        term.new_term_button
> +    Populate Field                          Term Name               Robot Automation Term 
> +    Click on Element                        term.account
> +    Press Keys                              //input[@title='Search Accounts']          ARROW_DOWN+RETURN
> +
> +
> +    Click on Element                        term.save_button
> +#    Click Modal Button                      Save
> ⬇️ Suggested change
> -#    Click Modal Button                      Save
> In robot/EDA/tests/browser/regression/run_regression_suite1.robot:
> 
> > +#    Click Modal Button                      Save
> +    Click on Element                        term.search_terms
> +    Click on Element                        term.new_term_button
> +    Populate Field                          Term Name               Robot Automation Term 
> +    Click on Element                        term.account
> +    Press Keys                              //input[@title='Search Accounts']          ARROW_DOWN+RETURN
> +
> +
> +    Click on Element                        term.save_button
> +#    Click Modal Button                      Save
> +    ${id_offering} =                        Generate Random String
> +
> +    Click on Element                        term.course_offering_id
> +    Input Text                              //span[contains(text(), 'Course Offering ID')]//../following-sibling::input         ${id_offering}
> +    Click on Element                        course_offering.next_save_button
> +#    Click Modal Button                      Save
> ⬇️ Suggested change
> -#    Click Modal Button                      Save
> In robot/EDA/tests/browser/settings/eda_settings_naming.robot:
> 
> > +
> +    Go To Eda Settings
> +
> +    Wait for Locator                            eda_settings.system_tab
> +    Click on Element                            eda_settings.system_tab
> +
> +
> +    # The following call to Select Checkbox In Eda Settings goes into EDIT mode, then SAVE the setting
> +    # So this call must NOT be called while in EDIT mode, else the Edit button will be not found.
> +    # This call places a checkmark in the Automatic Household Naming checkbox
> +    ${hh_naming_check} =   Get Eda Locator      eda_settings.hh_naming_check
> +    ${hh_role_checkbox} =  Get Eda Locator      eda_settings.hh_naming_role_checkbox
> +    Select Checkbox In Eda Settings             ${hh_naming_check}        ${hh_role_checkbox}
> +
> +
> +#    #Go into Edit Mode, to set the Administrative Account Name Format
> Remove dead code.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

dashattuck commented on robot/EDA/tests/browser/settings/eda_settings_naming.robot:
```diff
@@ -0,0 +1,35 @@
+    Select Checkbox In Eda Settings             ${hh_naming_check}        ${hh_role_checkbox}
+
+
+#    #Go into Edit Mode, to set the Administrative Account Name Format
```
thank you

dashattuck commented on robot/EDA/tests/browser/contacts/populate_contacts.robot:
```diff
@@ -0,0 +1,100 @@
+*** Settings ***
```
for test scenario purposes

boakley commented:
I took a look at the test failure, and it appears that the test is failing to open the app launcher. It appears that this is because there is a modal dialog open, possibly from a previous step.

A simple fix might be for the test to start by refreshing the page or explicitly going to the home page so that it is in a known state, rather than just picking up where a previous test left off.

dashattuck commented on robot/EDA/resources/EDA.robot:
```diff
@@ -43,6 +43,29 @@ API Create Course Enrollment
     &{enrollment} =   Salesforce Get  Course_Enrollment__c  ${enrollment_id}
     [return]          &{enrollment}
 
+API Populate Create And Return Contact with Address
```
Modified to remove API

dashattuck commented on robot/EDA/resources/EDA.robot:
```diff
@@ -43,6 +43,29 @@ API Create Course Enrollment
     [return]          &{enrollment}
 
+API Populate Create And Return Contact with Address
+    [Arguments]      ${first_name}      ${last_name}      ${mailing_street}      ${mailing_city}      ${mailing_zip}      ${mailing_state}      ${mailing_country}
```
split this over a few lines

dashattuck commented on robot/EDA/tests/browser/settings/eda_settings_accounts_and_contacts_edit_mode.robot:
```diff
@@ -0,0 +1,110 @@
+    
+    #Click Button                    Save
+
+    Sleep                           5
```
Removed the pauses

dashattuck commented on robot/EDA/tests/browser/settings/eda_settings_accounts_and_contacts_edit_mode.robot:
```diff
@@ -0,0 +1,110 @@
+
+
+    # Uncheck everything
+    Click Element                   //div[contains(@class,'slds-form-element__control')]//span[contains(text(),'Academic Program')]
```
Moved to locators file

dashattuck commented on robot/EDA/tests/browser/regression/run_regression_suite1.robot:
```diff
@@ -0,0 +1,157 @@
+    Click on Element                        new_program_enrollment_save_button
+
+    #Verify that we have ONE affiliated account
+    Reload Page
```
@boakley - this was per your suggestion to do a page refresh.  Is there a different way that should be used?

dashattuck replied:
need the reload to refresh and show the updated count

dashattuck commented on robot/EDA/tests/browser/regression/run_regression_suite1.robot:
```diff
@@ -0,0 +1,157 @@
+
+*** Test Cases ***
+    
+Create Contact And Verify
```
This is an actual test case

boakley commented on robot/EDA/tests/browser/settings/eda_settings_accounts_and_contacts_edit_mode.robot:
```diff
@@ -0,0 +1,126 @@
+    Click on Element                account_types.administrative
+
+    Wait for Locator                account_types.academic
+    Click on Element                account_types.academic
```
You could cut the size of this test almost in half and also make it easier to read if you moved the `Wait for Locator` inside the `Click on Element` keyword.

dashattuck replied:
Will do in round 2

dashattuck commented:
ah - yes - I wanted to figure out a way to put zip code back in.... so I
kept it there, but commented out because it kept failing on zip code.

And I'll remove that API and dead code.

Doug

On Fri, Jul 12, 2019 at 9:04 AM Bryan Oakley <notifications@github.com>
wrote:

> *@boakley* approved this pull request.
>
> There are a couple of things that I think need to be addressed, but none
> of them are showstoppers. I don't want to be a bottleneck for your team so
> I'm going to approve this PR. Please consider addressing these issues in
> the future.
> ------------------------------
>
> In robot/EDA/resources/EDA.robot
> <https://github.com/SalesforceFoundation/EDA/pull/770#discussion_r303016936>
> :
>
> > @@ -43,6 +43,29 @@ API Create Course Enrollment
>      &{enrollment} =   Salesforce Get  Course_Enrollment__c  ${enrollment_id}
>      [return]          &{enrollment}
>
> +API Populate Create And Return Contact with Address
>
> This keyword still has "API" in its name, but it doesn't use the API to
> create the contact. Is there a reason it's not using the API? The keyword
> shouldn't be named "API..." if it doesn't use the API.
> ------------------------------
>
> In robot/EDA/resources/locator.py
> <https://github.com/SalesforceFoundation/EDA/pull/770#discussion_r303018205>
> :
>
> > @@ -5,7 +5,8 @@
>          "button": "//div[@class='actionsContainer']/button[@title='{}']",
>          "datepicker": "//div[contains(@class,'uiDatePickerGrid')]/table[@class='calGrid']//span[text()='{}']",
>          "edit_button": '//*[@title="{}"]',
> -        "list": "//div[contains(@class,'forcePageBlockSectionRow')]/div[contains(@class,'forcePageBlockItem')]/div[contains(@class,'slds-p-vertical_xx-small')]/div[@class='slds-form-element__control']/div[.//span[text()='{}']][//div[contains(@class,'uiMenu')]//a[@class='select']]",
> +        "list_old": "//div[contains(@class,'forcePageBlockSectionRow')]/div[contains(@class,'forcePageBlockItem')]/div[contains(@class,'slds-p-vertical_xx-small')]/div[@class='slds-form-element__control']/div[.//span[text()='{}']][//div[contains(@class,'uiMenu')]//a[@class='select']]",
>
> Is this locator being used anywhere? If not, there's no point in keeping
> it around.
> ------------------------------
>
> In robot/EDA/resources/test_data.robot
> <https://github.com/SalesforceFoundation/EDA/pull/770#discussion_r303019292>
> :
>
> > +Library        EDA.py
> +
> +*** Keywords ***
> +Create Test User
> +    [Documentation]  This creates a test user via the salesforce API
> +    [Arguments]
> +    ...  ${first_name}  ${last_name}  ${mailing_street}  ${mailing_city}  ${mailing_state}  ${mailing_country}
> +
> +#${mailing_zipcode}
> +
> +    ${contact_id} =   Salesforce Insert  Contact
> +    ...                 FirstName=${first_name}
> +    ...                 LastName=${last_name}
> +    ...                 MailingStreet=${mailing_street}
> +    ...                 MailingCity=${mailing_city}
> +    #...                 MailingZipcode=${mailing_zipcode}
>
> What's the reason for this being commented out? Can you just remove the
> line entirely? The same goes for Primary_Address_Type__c a few lines
> further down.
> ------------------------------
>
> In robot/EDA/resources/test_data.robot
> <https://github.com/SalesforceFoundation/EDA/pull/770#discussion_r303019533>
> :
>
> > +    ${contact_id} =   Salesforce Insert  Contact
> +    ...                 FirstName=${first_name}
> +    ...                 LastName=${last_name}
> +    ...                 MailingStreet=${mailing_street}
> +    ...                 MailingCity=${mailing_city}
> +    #...                 MailingZipcode=${mailing_zipcode}
> +    ...                 MailingState=${mailing_state}
> +    ...                 MailingCountry=${mailing_country}
> +
> +    #...                 Primary_Address_Type__c=Work
> +
> +
> +Initialize Test Data
> +
> +    Open Test Browser
> +    #                 First    Last       Street                 City         Zip       State  Country
>
> What's the point in having all of these tests commented out? Is there a
> reason you don't just delete them?
> ------------------------------
>
> In robot/EDA/tests/browser/contacts/populate_contacts.robot
> <https://github.com/SalesforceFoundation/EDA/pull/770#discussion_r303020467>
> :
>
> > @@ -0,0 +1,100 @@
> +*** Settings ***
> +
> +Resource        robot/EDA/resources/EDA.robot
> +Suite Setup     Open Test Browser
> +Suite Teardown  Delete Records and Close Browser
> +
> +*** Test Cases ***
> +
> +
> +
> +Let Us Populate Create And Return Contact with Address
>
> Can't all of these tests be deleted? Aren't these just doing the same
> thing that the Initialize Test Data keyword is doing?
> ------------------------------
>
> In
> robot/EDA/tests/browser/settings/eda_settings_accounts_and_contacts_edit_mode.robot
> <https://github.com/SalesforceFoundation/EDA/pull/770#discussion_r303022394>
> :
>
> > +
> +Validate Edit Mode For Accounts and Contacts, Settings
> +    [tags]  unstable
> +    Go To Eda Settings
> +    Wait for Locator            	tabs.accountsandcontacts
> +    Click on Element                tabs.accountsandcontacts
> +
> +    #Go into Edit mode
> +    #Click Button                    Edit
> +    Click on Element                account_types.edit
> +
> +    # Select Multi-Address Account Types
> +    Click on Element                account_types.administrative
> +
> +    Wait for Locator                account_types.academic
> +    Click on Element                account_types.academic
>
> You could cut the size of this test almost in half and also make it easier
> to read if you moved the Wait for Locator inside the Click on Element
> keyword.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/SalesforceFoundation/EDA/pull/770?email_source=notifications&email_token=AH77T2BWUL5WQVUPHZEINKDP7CMQPA5CNFSM4HH3L45KYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB6JS7VI#pullrequestreview-261304277>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AH77T2DGFKP3ZSLT7L5KUNLP7CMQPANCNFSM4HH3L45A>
> .
>


-- 
Doug Shattuck
Test Automation Engineer | Salesforce
Mobile: 385-242-9732 <(385)%20242-9732>


<http://smart.salesforce.com/sig/dshattuck//us_mb_kb/default/link.html>

dashattuck commented on robot/EDA/resources/test_data.robot:
```diff
@@ -0,0 +1,51 @@
+Initialize Test Data
+
+    Open Test Browser
+    #                 First    Last       Street                 City         Zip       State  Country
```
removed

dashattuck commented on robot/EDA/resources/test_data.robot:
```diff
@@ -0,0 +1,51 @@
+    ...                 LastName=${last_name}
+    ...                 MailingStreet=${mailing_street}
+    ...                 MailingCity=${mailing_city}
+    #...                 MailingZipcode=${mailing_zipcode}
```
I removed those two lines.... they were placeholders, but it's okay for me to remove them.

dashattuck commented on robot/EDA/tests/browser/contacts/populate_contacts.robot:
```diff
@@ -0,0 +1,100 @@
+
+    
+    
+Let Us Populate Create And Return Contact with Address
```
They are validating the process of calling them two different ways

dashattuck commented on robot/EDA/resources/EDA.robot:
```diff
@@ -43,6 +43,29 @@ API Create Course Enrollment
     &{enrollment} =   Salesforce Get  Course_Enrollment__c  ${enrollment_id}
     [return]          &{enrollment}
 
+API Populate Create And Return Contact with Address
```
Removed the API string

dashattuck commented on robot/EDA/resources/locator.py:
```diff
@@ -5,7 +5,8 @@
         "datepicker": "//div[contains(@class,'uiDatePickerGrid')]/table[@class='calGrid']//span[text()='{}']",
         "edit_button": '//*[@title="{}"]',
-        "list": "//div[contains(@class,'forcePageBlockSectionRow')]/div[contains(@class,'forcePageBlockItem')]/div[contains(@class,'slds-p-vertical_xx-small')]/div[@class='slds-form-element__control']/div[.//span[text()='{}']][//div[contains(@class,'uiMenu')]//a[@class='select']]",
+        "list_old": "//div[contains(@class,'forcePageBlockSectionRow')]/div[contains(@class,'forcePageBlockItem')]/div[contains(@class,'slds-p-vertical_xx-small')]/div[@class='slds-form-element__control']/div[.//span[text()='{}']][//div[contains(@class,'uiMenu')]//a[@class='select']]",
```
It's still useful - I go back and forth while developing and debugging

robertbwatson commented:
@juliesnow  This is the PR that we were confused about on the release notes this morning. Do we usually have release notes for Robot test changes? If so, maybe we just need something like "We've added several new test automation scripts." ?

davisagli commented:
@robertbwatson we don't include notices about changes to our test automation in release notes, since they are not customer-facing. This PR description can be edited to move those comments up above the "Critical Changes" heading so they aren't included in release notes.

juliesnow commented:
@davisagli - Not following, are you saying that text we include in Critical Changes won't be included in RNs? We've not needed to include notes there since I've been on the team, but it was my understanding that section is where we'd convey really critical info that's going out in a release. Alternatively, could we just delete the text from "Changes?"

davisagli commented:
@juliesnow no, I'm saying text *above* Critical Changes (i.e. before any heading) will not be included in the release notes.

juliesnow commented:
@robertbwatson - Adding the text you suggested above the heading and removing the other from Changes.

jacebryan commented:
CCE feedback on release note usability:

"We've added several new test automation scripts."

- What new test automation scripts were introduced?
- What do these new test automation scripts do?
- Why were these new test automation scripts introduced?
- Who would find the most use when it comes to these new test automation scripts?
